atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/invariant/SnapshotInvariantTest.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.workload.invariant;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.workload.store.ReadableTransactionStore;
+import com.palantir.atlasdb.workload.transaction.WitnessedTransactionsBuilder;
+import com.palantir.atlasdb.workload.transaction.WorkloadTestHelpers;
+import com.palantir.atlasdb.workload.transaction.witnessed.InvalidWitnessedTransaction;
+import com.palantir.atlasdb.workload.transaction.witnessed.InvalidWitnessedTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransaction;
+import com.palantir.atlasdb.workload.workflow.ImmutableWorkflowHistory;
+import com.palantir.atlasdb.workload.workflow.WorkflowHistory;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public final class SnapshotInvariantTest {
```

Comment:
In general the coverage here is solid, though I'd recommend validating the "elementary" property that we check for failures/errors in reading from the correct snapshot version - e.g. write v1 (start 10 commit 20), start 15 commit 25 = empty, start 22 commit 25 = v1.